File: code/string_algorithms/src/longest_palindromic_substring.java
// Java implementation of O(n^2)
// time and O(1) space method
// to find the longest palindromic substring
public class LongestPalinSubstring {
	// A utility function to print
	// a substring str[low..high]
	static void printSubStr(String str,
							int low, int high)
	{
		System.out.println(
			str.substring(
				low, high + 1));
	}

	// This function prints the
	// longest palindrome substring
	// (LPS) of str[]. It also
	// returns the length of the
	// longest palindrome
	static int longestPalSubstr(String str)
	{
		// The result (length of LPS)
		int maxLength = 1;

		int start = 0;
		int len = str.length();

		int low, high;

		// One by one consider every
		// character as center
		// point of even and length
		// palindromes
		for (int i = 1; i < len; ++i) {
			// Find the longest even
			// length palindrome with
			// center points as i-1 and i.
			low = i - 1;
			high = i;
			while (low >= 0 && high < len
				&& str.charAt(low)
						== str.charAt(high)) {
				--low;
				++high;
			}
		
			// Move back to the last possible valid palindrom substring
			// as that will anyway be the longest from above loop
			++low; --high;
			if (str.charAt(low) == str.charAt(high) && high - low + 1 > maxLength) {
				start = low;
				maxLength = high - low + 1;
			}

			// Find the longest odd length
			// palindrome with center point as i
			low = i - 1;
			high = i + 1;
			while (low >= 0 && high < len
				&& str.charAt(low)
						== str.charAt(high)) {
				--low;
				++high;
			}
		
			// Move back to the last possible valid palindrom substring
			// as that will anyway be the longest from above loop
			++low; --high;
			if (str.charAt(low) == str.charAt(high) && high - low + 1 > maxLength) {
				start = low;
				maxLength = high - low + 1;
			}
		}

		System.out.print("Longest palindrome substring is: ");
		printSubStr(str, start, start + maxLength - 1);

		return maxLength;
	}

	// Driver program to test above function
	public static void main(String[] args)
	{

		String str = "todayisabeautifulday";
		System.out.println("Length is: " + longestPalSubstr(str));
	}
}

File: code/string_algorithms/src/longest_palindromic_substring/longest_palindromic_substring.cpp
// Part of Cosmos by OpenGenus
#include <iostream>
#include <string>
using namespace std;

// expand in both directions of low and high to find
// maximum length palindrome
string expand(string str, int low, int high)
{
	int len = str.length();

	// run till str[low.high] is a palindrome
	while (low >= 0 && high < len && (str[low] == str[high]))
		low--, high++;		// expand in both directions

	// return palindromic substring
	return str.substr(low + 1, high - low - 1);
}

// Function to find Longest Palindromic Substring in O(n^2) time
// and O(1) space
string LongestPalindromicSubstring(string str, int len)
{
	// max_str stores the maximum length palindromic substring
	// found so far

	string max_str = "", curr_str;

	// max_length stores the length of maximum length palindromic
	// substring found so far

	int max_length = 0, curr_length;

	// consider every character of given string as mid points and expand
	// in its both directions to find maximum length palindrome

	for (int i = 0; i < len; i++)
	{
		// find a longest odd length palindrome with str[i] as mid point

		curr_str = expand(str, i, i);
		curr_length = curr_str.length();

		// update maximum length palindromic substring if odd length
		// palindrome has greater length

		if (curr_length > max_length)
		{
			max_length = curr_length;
			max_str = curr_str;
		}

		// find a longest even length palindrome with str[i] and str[i+1]
		// as mid points. Note that an even length palindrome has two
		// mid points

		curr_str = expand(str, i, i + 1);
		curr_length = curr_str.length();
	
		// update maximum length palindromic substring if even length
		// palindrome has greater length

		if (curr_length > max_length)
		{
			max_length = curr_length;
			max_str = curr_str;
		}
	}

	return max_str;
}

// main function
int main()
{
	string str;
	cin>>str;

	cout << "Longest Palindromic Substring of " << str << " is "
			<< LongestPalindromicSubstring(str, str.length() - 1);

	return 0;
}
